buffer/src/main/java/io/netty/buffer/UnpooledByteBufAllocator.java

```diff
@@ -85,4 +95,126 @@ public CompositeByteBuf compositeDirectBuffer(int maxNumComponents) {
     public boolean isDirectBufferPooled() {
         return false;
     }
+
+    @Override
+    public long usedHeapMemory() {
+        return heapCounter.value();
+    }
+
+    @Override
+    public long usedDirectMemory() {
+        return directCounter.value();
+    }
+
+    private static final class TrackedUnpooledUnsafeHeapByteBuf extends UnpooledUnsafeHeapByteBuf {
+        TrackedUnpooledUnsafeHeapByteBuf(UnpooledByteBufAllocator alloc, int initialCapacity, int maxCapacity) {
+            super(alloc, initialCapacity, maxCapacity);
+        }
+
+        @Override
+        byte[] allocateArray(int initialCapacity) {
+            byte[] bytes = super.allocateArray(initialCapacity);
+            ((UnpooledByteBufAllocator) alloc()).heapCounter.add(bytes.length);
+            return bytes;
+        }
+
+        @Override
+        void freeArray(byte[] array) {
+            int length = array.length;
+            super.freeArray(array);
+            ((UnpooledByteBufAllocator) alloc()).heapCounter.add(-length);
+        }
+    }
+
+    private static final class TrackedUnpooledHeapByteBuf extends UnpooledHeapByteBuf {
+        TrackedUnpooledHeapByteBuf(UnpooledByteBufAllocator alloc, int initialCapacity, int maxCapacity) {
+            super(alloc, initialCapacity, maxCapacity);
+        }
+
+        @Override
+        byte[] allocateArray(int initialCapacity) {
+            byte[] bytes = super.allocateArray(initialCapacity);
+            ((UnpooledByteBufAllocator) alloc()).heapCounter.add(bytes.length);
+            return bytes;
+        }
+
+        @Override
+        void freeArray(byte[] array) {
+            int length = array.length;
+            super.freeArray(array);
+            ((UnpooledByteBufAllocator) alloc()).heapCounter.add(-length);
+        }
+    }
+
+    private static final class TrackedUnpooledUnsafeNoCleanerDirectByteBuf
+            extends UnpooledUnsafeNoCleanerDirectByteBuf {
+        TrackedUnpooledUnsafeNoCleanerDirectByteBuf(
+                UnpooledByteBufAllocator alloc, int initialCapacity, int maxCapacity) {
+            super(alloc, initialCapacity, maxCapacity);
+        }
+
+        @Override
+        protected ByteBuffer allocateDirect(int initialCapacity) {
+            ByteBuffer buffer = super.allocateDirect(initialCapacity);
+            ((UnpooledByteBufAllocator) alloc()).directCounter.add(buffer.capacity());
+            return buffer;
+        }
+
+        @Override
+        ByteBuffer reallocateDirect(ByteBuffer oldBuffer, int initialCapacity) {
+            int capacity = oldBuffer.capacity();
+            ByteBuffer buffer = super.reallocateDirect(oldBuffer, initialCapacity);
+            ((UnpooledByteBufAllocator) alloc()).directCounter.add(buffer.capacity() - capacity);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Cast one of the operands of this substraction operation to a "long". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2184)